R/chat-async.R

```diff
@@ -0,0 +1,271 @@
+#' ") %...>% cat()
+#'
+#' chat <- new_chat_async_openai()
+#' chat$register_tool(
+#'   fun = rnorm,
+#'   name = "rnorm",
+#'   description = "Drawn numbers from a random normal distribution",
+#'   arguments = list(
+#'     n = tool_arg(
+#'       type = "integer",
+#'       description = "The number of observations. Must be a positive integer."
+#'     ),
+#'     mean = tool_arg(
+#'       type = "number",
+#'       description = "The mean value of the distribution."
+#'     ),
+#'     sd = tool_arg(
+#'       type = "number",
+#'       description = "The standard deviation of the distribution. Must be a non-negative number."
+#'     )
+#'   )
+#' )
+#' chat$chat("
+#'   Give me five numbers from a random normal distribution.
+#'   Briefly explain your work.
+#' ") %...>% cat()
+new_chat_async_openai <- function(system_prompt = NULL,
+                                  base_url = "https://api.openai.com/v1",
+                                  api_key = openai_key(),
+                                  model = "gpt-4o-mini",
+                                  quiet = FALSE) {
+  check_string(system_prompt, allow_null = TRUE)
+  check_string(base_url)
+  check_string(api_key)
+  check_string(model)
+  check_bool(quiet)
+
+  system_prompt <- system_prompt %||%
+    "You are a helpful assistant from New Zealand who is an experienced R programmer"
+
+  ChatAsyncOpenAI$new(
+    base_url = base_url,
+    model = model,
+    api_key = api_key,
+    system_prompt = system_prompt,
+    quiet = quiet
+  )
+}
+
+#' @rdname new_chat_async_openai
+ChatAsyncOpenAI <- R6::R6Class("ChatOpenAI",
+  public = list(
+    #' @param system_prompt A system prompt to set the behavior of the assistant.
+    #' @param base_url The base URL to the endpoint; the default uses ChatGPT.
+    #' @param api_key The API key to use for authentication. You generally should
+    #'   not supply this directly, but instead set the `OPENAI_API_KEY` environment
+    #'   variable.
+    #' @param model The model to use for the chat; defaults to GPT-4o mini.
+    #' @param quiet If `TRUE` does not print output as its received.
+    initialize = function(base_url, model, api_key, system_prompt, quiet = TRUE) {
+      private$base_url <- base_url
+      private$model <- model
+      private$api_key <- api_key
+      private$quiet <- quiet
+
+      private$add_message(list(
+        role = "system",
+        content = system_prompt
+      ))
+    },
+
+    #' @description Submit text to the chatbot, and receive a promise that
+    #'   resolves with the response all at once.
+    #' @param text The text to send to the chatbot.
+    #' @returns A promise that resolves to a string (probably Markdown).
+    chat = function(text) {
+      check_string(text)
+
+      # Returns a single message (the final response from the assistant), even if
+      # multiple rounds of back and forth happened.
+      done <- coro::async_collect(private$chat_impl(text, stream = FALSE))
```

Comment:
I guess this is a fundamental part of async, but it kind of sucks that it has to infect the whole stack, meaning that we need two versions of every method and function (sync and async).

Reply:
Yes. This is known as the "what color is your function?" problem. I spent some time considering writing a meta-version of each such method, then being able to "project" sync and async versions of them. It might make sense at some point, but for now it seemed like the bar for contributing to this package would go sky high. (Already higher than I'd like with async, period...)